### PR TITLE
Lvruby 14 2 Added filtering, pagination and search functionality to news.

### PR DESCRIPTION
**Pull request by nataliakozoriz**
## JIRA ticket

_Please provide the URL for the JIRA ticket._

## Code reviewers

- [x] @omaik 
- [x] @kmandziy 

## Summary of issue

Add  news filtration via category, add search via news, add pagination.

## Summary of change

Created partials which provide required functions.

## Review checklist

- [x] Rubocop doesn't raise any warnings and offenses
- [x] New features are covered with new specs
- [x] All specs pass
Screenshot from 2019-05-06 13-44-25

### CODE REVIEW

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.04%** covered at **3.34** hits/line

File: app/views/news/_form.html.haml [hunk 1]
```diff
@@ -1,4 +1,3 @@
-= tinymce_assets
```
Comment (okohuch) [inline app/views/news/_form.html.haml]: why did you remove this?

File: app/models/news.rb [hunk 2]
```diff
@@ -2,8 +2,17 @@
 
 class News < ApplicationRecord
   belongs_to :user
+  belongs_to :category, optional: true
 
   validates :title, presence: true, length: { maximum: 100 }
   validates :body, presence: true
   validates :short_information, presence: true, length: { maximum: 200 }
+  validates :category_id, presence: true
```
Comment (okohuch) [inline app/models/news.rb]: you added `optional: true` but here you still validate `presence`.
remove `optional: true`. 
`belongs_to` will require parent object by default, therefore, do not think we need this validation.

**Comment by okohuch** [comment]
@nataliakozoriz please add screenshots

**Comment by nataliakozoriz** [comment]
> @nataliakozoriz please add screenshots

Attached screen.

File: app/views/news/_form.html.haml (hunk 1 above)
Comment (nataliakozoriz) [reply app/views/news/_form.html.haml]: That was accidentally, fixed it.

File: app/models/news.rb (hunk 2 above)
Comment (nataliakozoriz) [reply app/models/news.rb]: Deleted optional: true.

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.0%** covered at **3.4** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.04%** covered at **3.38** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.08%** covered at **3.43** hits/line

File: app/assets/stylesheets/pages/news.scss [hunk 3]
```diff
@@ -54,3 +54,13 @@ img .portrait {
 .img-comment {
   width: 70px;
 }
+
+.pagination {
+  margin-top: 20px;
+
+  a,
```
[the remark this answers is not in the document]
Comment (omaik) [reply app/assets/stylesheets/pages/news.scss]: please do not mark the comment as resolved when it doesn't

File: app/controllers/news_controller.rb [hunk 4]
```diff
@@ -2,9 +2,13 @@
...
 class NewsController < ApplicationController
   before_action :find_current_user_news, only: %i[edit update destroy]
+  has_scope :by_categories, type: :array
 
   def index
-    @news = News.all
+    @categories = Category.all
+    search = params[:search]
+    @news = apply_scopes(News).all.search(search).paginate(page: params[:page], per_page: 3)
+    redirect_to news_index_path, alert: t('.alert') if @news.count.zero?
```
Comment (omaik) [inline app/controllers/news_controller.rb]: this redirect seems redundant, you could just add `flash.now[:alert]` and it should behave in the same way without making additional request

File: app/assets/stylesheets/pages/news.scss (hunk 3 above)
[the remark this answers is not in the document]
Comment (nataliakozoriz) [reply app/assets/stylesheets/pages/news.scss]: I can't change it in case previous_page class is defined at gem specification.
Screenshot from 2019-05-08 12-58-17

File: app/controllers/news_controller.rb (hunk 4 above)
Comment (nataliakozoriz) [reply app/controllers/news_controller.rb]: If I change controller code to _flash.now[:alert] = t('.alert') if @news.count.zero?_ the results is:
Screenshot from 2019-05-08 13-17-48
And my idea was to display message : Nothing was found and display all news.

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.08%** covered at **3.43** hits/line

File: app/controllers/news_controller.rb (hunk 4 above)
Comment (nataliakozoriz) [reply app/controllers/news_controller.rb]: Fixed. Now filters are not clearing while search found nothing.
Screenshot from 2019-05-09 14-37-10

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **96.08%** covered at **3.43** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.08%** covered at **5.16** hits/line

File: app/views/news/_selector.html.haml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+          category.id,
+          params.key?(:by_categories) ? params[:by_categories].include?(category.id.to_s) : false,
+          id: category.name, class: 'box'
+        = label_tag category.name
+    = button_tag( t('news.selector.button'), class: "btn btn-primary")
+    %br
+    .col-sm-12
+      .search
+        %i.fa.fa-search{ "aria-hidden" => "true" }
+        = text_field_tag  :search,
```
Comment (omaik) [inline app/views/news/_selector.html.haml]: please name the field title, search is too abstract

File: app/models/news.rb [hunk 6]
```diff
@@ -2,8 +2,17 @@
...
   belongs_to :user
+  belongs_to :category
 
   validates :title, presence: true, length: { maximum: 100 }
   validates :body, presence: true
   validates :short_information, presence: true, length: { maximum: 200 }
+  validates :category_id, presence: true
+
+  scope :by_categories, ->(category) { where(category: category) }
+  default_scope { order(created_at: :desc) }
```
Comment (omaik) [inline app/models/news.rb]: if all your queries order by created_at it is probably a nice idea to index field

File: app/views/news/_selector.html.haml (hunk 5 above)
Comment (nataliakozoriz) [reply app/views/news/_selector.html.haml]: changed to news_search

Comment (omaik) [reply app/views/news/_selector.html.haml]: You searching by title only, so it does make sense to name it accordingly. It is hard to understand without reading of several classes what `news_search` means

Comment (nataliakozoriz) [reply app/views/news/_selector.html.haml]: Changed to news_title_search.

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.08%** covered at **5.16** hits/line

File: app/models/news.rb (hunk 6 above)
Comment (nataliakozoriz) [reply app/models/news.rb]: Added index to created_at

File: app/views/news/_selector.html.haml (hunk 5 above)
Comment (omaik) [reply app/views/news/_selector.html.haml]: I still prefer just `title` due to its simplicity, but if you insist let it be.

File: app/controllers/news_controller.rb [hunk 7]
```diff
@@ -2,9 +2,13 @@
 
 class NewsController < ApplicationController
   before_action :find_current_user_news, only: %i[edit update destroy]
+  has_scope :by_categories, type: :array
 
   def index
-    @news = News.all
+    @categories = Category.all
+    search = params[:search]
```
Comment (omaik) [inline app/controllers/news_controller.rb]: you renamed search field, but still have just search in params

File: app/models/news.rb [hunk 8]
```diff
@@ -2,8 +2,17 @@
...
+
+  scope :search, lambda { |search|
```
Comment (omaik) [inline app/models/news.rb]: please rename it accordingly to what it does, f.e. `search_by_title`

File: app/views/news/_selector.html.haml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    .col-sm-12
+      .search
+        %i.fa.fa-search{ "aria-hidden" => "true" }
+        = text_field_tag  :news_title_search,
+          params[:search],
+          placeholder:  t('news.search.field'),
+          class: "form-control"
+
+
+
```
Comment (omaik) [inline app/views/news/_selector.html.haml]: too much empty lines in the end.

File: app/models/news.rb (hunk 8 above)
Comment (nataliakozoriz) [reply app/models/news.rb]: renaned to title_search

File: app/controllers/news_controller.rb (hunk 7 above)
Comment (nataliakozoriz) [reply app/controllers/news_controller.rb]: Renamed to title_search

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.08%** covered at **5.16** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.49%** covered at **4.98** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.49%** covered at **4.98** hits/line

File: app/models/news.rb (hunk 2 above)
Comment (nataliakozoriz) [reply app/models/news.rb]: Deleted redundant validation.

File: app/controllers/news_controller.rb [hunk 10]
```diff
@@ -2,9 +2,13 @@
...
 class NewsController < ApplicationController
   before_action :find_current_user_news, only: %i[edit update destroy]
+  has_scope :by_categories, type: :array
 
   def index
-    @news = News.all
+    @categories = Category.all
+    search = params[:title_search]
+    @news = apply_scopes(News).all.title_search(search).paginate(page: params[:page], per_page: 3)
+    flash.now[:alert] = t('.alert') if @news.count.zero?
```
[the remark this answers is not in the document]
Comment (nataliakozoriz) [reply app/controllers/news_controller.rb]: Changed to @news.empty?

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.48%** covered at **5.0** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.5%** covered at **4.96** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.48%** covered at **4.95** hits/line

**Comment by vikabelous** [comment]
**Code Coverage:**
:green_heart: **97.48%** covered at **4.95** hits/line
